apps/openssl.c

```diff
@@ -285,17 +287,24 @@ int main(int argc, char *argv[])
         global_help = argc > 1
             && (strcmp(argv[1], "-help") == 0 || strcmp(argv[1], "--help") == 0
                 || strcmp(argv[1], "-h") == 0 || strcmp(argv[1], "--h") == 0);
+        global_version = argc > 1
+            && (strcmp(argv[1], "-version") == 0 || strcmp(argv[1], "--version") == 0
+                || strcmp(argv[1], "-v") == 0 || strcmp(argv[1], "--v") == 0);
+
         argc--;
         argv++;
-        opt_appname(argc == 1 || global_help ? "help" : argv[0]);
+        opt_appname(argc == 1 || global_help ? "help" : global_version ? "version" : argv[0]);
```

Comment:
I left _unchanged_ the `argc == 1` portion, but this is semi-confusing to me.:

- If the user just typed `openssl` then `argc` is 1 at line [287](https://github.com/openssl/openssl/pull/20936/files#diff-1256ca7634563671d3168d89b2804241a92a693a2f8524cc8fde016d6e3bdf20L285) thus that stuff doesn't fire, nor does the new code at line ~x~/290 we then decrement the `argc` and bump up `argv` at line [~288-289~/294-295](https://github.com/openssl/openssl/pull/20936/files#diff-1256ca7634563671d3168d89b2804241a92a693a2f8524cc8fde016d6e3bdf20L288-L289) so `argc == 0` and `argv` points to the next (non-existent) argument! Thus line [~290~/296](https://github.com/openssl/openssl/pull/20936/files#diff-1256ca7634563671d3168d89b2804241a92a693a2f8524cc8fde016d6e3bdf20L290-R296) will pass `argv[0]` (which is an index-past-bounds) even though there is no command in that position to `opt_appname()`. I assume it should really be treating this as a help command (which it indeed does thanks to the ternary at line [~296-297~/303-304](https://github.com/openssl/openssl/pull/20936/files#diff-1256ca7634563671d3168d89b2804241a92a693a2f8524cc8fde016d6e3bdf20L295-R307)). 
- If the user if the user just typed `openssl --help` (existing behavior) or `openssl --version` (new behavior), then `argc == 2` and `argv[1]` has the specified option. Then the code at lines [~287-289~/287-292](https://github.com/openssl/openssl/pull/20936/files#diff-1256ca7634563671d3168d89b2804241a92a693a2f8524cc8fde016d6e3bdf20L287-R292) will fire appropriately, setting the flag for the appropriate mode flag. Then the decrement and bump happens and now `argc == 1` and `argv[0]` points at the option. This means that we also pass the hard-coded `"help"` string through, instead of the (likely more correct) hard-coded `"version"` string to `opt_appname()` so the output would not actually be correct if the ternary magic did not run.

Please correct my understanding or let me know if we should be tweaking this a bit more.... I suspect the minimal change would be to remove the `argc == 1 ||` portion from the new code, and also make sure we do the right thing if (decremented) `argc == 0` (e.g. pass a null, or the `pname` extracted above). 
```c
opt_appname(global_help ? "help" : global_version ? "version" : arcg >= 1 ? argv[0] : pname);
```

We could be even simpler if we do the `opt_appname()` call inside a less-cleaver if / else if / else block than the (newly nested) ternary at lines [~295-298~/301-307](https://github.com/openssl/openssl/pull/20936/files#diff-1256ca7634563671d3168d89b2804241a92a693a2f8524cc8fde016d6e3bdf20L295-R307), but I am not sure that much change is desired.

My biggest question is really if I am reading this right in the zero-arguments case and the argv[0] access at old line [290](https://github.com/openssl/openssl/pull/20936/files#diff-1256ca7634563671d3168d89b2804241a92a693a2f8524cc8fde016d6e3bdf20L290) is actually beyond boundary?